service/history/historyEngine_test.go

```diff
@@ -1666,6 +1666,82 @@ func (s *engineSuite) TestRespondWorkflowTaskCompletedSingleActivityScheduledWor
+}
+
+func (s *engineSuite) TestRespondWorkflowTaskCompleted_SkipSignalTaskGeneration() {
+	resp := s.testRespondWorkflowTaskCompletedSignalGeneration(true)
+	s.Nil(resp.GetStartedResponse())
+}
+
+func (s *engineSuite) testRespondWorkflowTaskCompletedSignalGeneration(skipGenerateTask bool) *historyservice.RespondWorkflowTaskCompletedResponse {
+	we := commonpb.WorkflowExecution{
+		WorkflowId: tests.WorkflowID,
+		RunId:      tests.RunID,
+	}
+	tl := "testTaskQueue"
+	tt := &tokenspb.Task{
+		Attempt:          1,
+		NamespaceId:      tests.NamespaceID.String(),
+		WorkflowId:       tests.WorkflowID,
+		RunId:            we.GetRunId(),
+		ScheduledEventId: 2,
+	}
+	taskToken, _ := tt.Marshal()
+	identity := "testIdentity"
+
+	signal := workflowservice.SignalWorkflowExecutionRequest{
+		Namespace:                tests.NamespaceID.String(),
+		WorkflowExecution:        &we,
+		Identity:                 identity,
+		SignalName:               "test signal name",
+		Input:                    payloads.EncodeString("test input"),
+		SkipGenerateWorkflowTask: skipGenerateTask,
+		RequestId:                uuid.New(),
+	}
+	signalRequest := &historyservice.SignalWorkflowExecutionRequest{
+		NamespaceId:   tests.NamespaceID.String(),
+		SignalRequest: &signal,
+	}
+
+	ms := workflow.TestLocalMutableState(s.mockHistoryEngine.shard, s.eventsCache,
+		tests.LocalNamespaceEntry, log.NewTestLogger(), we.GetRunId())
+	addWorkflowExecutionStartedEvent(ms, we, "wType", tl, payloads.EncodeString("input"), 100*time.Second, 90*time.Second, 200*time.Second, identity)
+	wt := addWorkflowTaskScheduledEvent(ms)
+	addWorkflowTaskStartedEvent(ms, wt.ScheduledEventID, tl, identity)
+
+	_, sigErr := ms.AddWorkflowExecutionSignaled(signal.GetSignalName(), signal.GetInput(), signal.GetIdentity(), signal.GetHeader(), signal.GetSkipGenerateWorkflowTask())
+	s.NoError(sigErr)
+	ms.AddSignalRequested(signal.GetRequestId())
+
+	wfMs := workflow.TestCloneToProto(ms)
+	gwmsResponse := &persistence.GetWorkflowExecutionResponse{State: wfMs}
+	s.mockExecutionMgr.EXPECT().GetWorkflowExecution(gomock.Any(), gomock.Any()).Return(gwmsResponse, nil)
+	s.mockExecutionMgr.EXPECT().UpdateWorkflowExecution(gomock.Any(), gomock.Any()).Return(tests.UpdateWorkflowExecutionResponse, nil)
+
+	_, err := s.mockHistoryEngine.SignalWorkflowExecution(context.Background(), signalRequest)
+	s.NoError(err)
+
+	var commands []*commandpb.Command
+	resp, err := s.mockHistoryEngine.RespondWorkflowTaskCompleted(context.Background(), &historyservice.RespondWorkflowTaskCompletedRequest{
+		NamespaceId: tests.NamespaceID.String(),
+		CompleteRequest: &workflowservice.RespondWorkflowTaskCompletedRequest{
+			TaskToken:             taskToken,
+			Commands:              commands,
+			Identity:              identity,
+			ReturnNewWorkflowTask: true,
+		},
+	})
+	s.NoError(err)
+	s.NotNil(resp)
+
+	s.Equal(enumsspb.WORKFLOW_EXECUTION_STATE_RUNNING, ms.GetExecutionState().State)
+
+	return resp
```

Comment:
I don't quite follow. 
- When handling the signal request, the logic will load mutable state and found the signal is already there I think (dedup based on requestID)?
- I don't see an assert for if a workflow task is generated?

Reply:
I think the point here is that the mutable state you created already contains the signal and SignalWorkflowExecution is again called for the same signal which makes it a no-op. 

So if you just want to the test the buffered event filtering logic, then I think there's no need for the separate signalWorkflowExecution call.